File: 2018/day23.py
import re
from runner import runner

class position:
    def __init__(self, x, y, z):
        self.x = x
        self.y = y
        self.z = z

    def distance_to(self, other):
        return abs(self.x - other.x) + abs(self.y - other.y) + abs(self.z - other.z)

    def distance_to_origin(self):
        return abs(self.x) + abs(self.y) + abs(self.z)

    def translate(self, dx, dy, dz):
        return position(self.x + dx, self.y + dy, self.z + dz)

class nanobot:
    def __init__(self, x, y, z, r):
        self.pos = position(x, y, z)
        self.r = r

    def in_range(self, pos):
        return self.pos.distance_to(pos) <= self.r

class cube:
    def __init__(self, center, r, bots):
        self.center = center
        self.min_pos = position(center.x - r, center.y - r, center.z - r)
        self.max_pos = position(center.x + r - 1, center.y + r - 1, center.z + r - 1)
        self.r = r
        self.bots = list(filter(lambda b: self.intersects(b), bots))
        self.bots_in_range = len(self.bots)

    def intersects(self, bot):
        if self.r == 0:
            return bot.in_range(self.center)
        # Calculate minimum distance required to travel along each axis to
        # arrive at the bounding box. Sum to find the Manhattan distance.
        distance = (
            self.min_axis_dist(bot.pos.x, self.min_pos.x, self.max_pos.x) +
            self.min_axis_dist(bot.pos.y, self.min_pos.y, self.max_pos.y) +
            self.min_axis_dist(bot.pos.z, self.min_pos.z, self.max_pos.z))
        return distance <= bot.r

    def min_axis_dist(self, target, a_min, a_max):
        if target < a_min:
            return a_min - target
        elif target > a_max:
            return target - a_max
        else:
            return 0

    def split(self):
        if self.r == 0:
            return []
        new_r = self.r // 2
        low = -new_r if new_r > 0 else -1
        high = new_r
        cubes = [
            cube(self.center.translate(low, low, low), new_r, self.bots),
            cube(self.center.translate(low, low, high), new_r, self.bots),
            cube(self.center.translate(low, high, low), new_r, self.bots),
            cube(self.center.translate(low, high, high), new_r, self.bots),
            cube(self.center.translate(high, low, low), new_r, self.bots),
            cube(self.center.translate(high, low, high), new_r, self.bots),
            cube(self.center.translate(high, high, low), new_r, self.bots),
            cube(self.center.translate(high, high, high), new_r, self.bots)
        ]
        cubes.sort(key = lambda c: c.bots_in_range, reverse = True)
        return cubes

class day23(runner):
    def __init__(self):
        self.bots = []

    def day(self):
        return 23

    def input(self, line):
        m = re.match(r'pos=<(-?\d+),(-?\d+),(-?\d+)>, r=(\d+)', line)
        self.bots.append(nanobot(int(m.group(1)), int(m.group(2)), int(m.group(3)), int(m.group(4))))

    def solve1(self):
        max_bot = max(self.bots, key = lambda b: b.r)
        in_range = sum(1 for b in self.bots if max_bot.in_range(b.pos))
        return str(in_range)

    def solve2(self):
        self.max_bots = 0
        self.max_position = position(0, 0, 0)
        start_cube = cube(position(0, 0, 0), 2**30, self.bots)
        self.binary_search([start_cube])
        return str(self.max_position.distance_to_origin())

    def binary_search(self, cubes):
        for c in cubes:
            if c.bots_in_range > self.max_bots:
                # print("Searching cube %d %d %d %d" % (c.center.x, c.center.y, c.center.z, c.r))
                self.update_search_state(c)
                self.binary_search(c.split())

    def update_search_state(self, c):
        if c.r != 0 or c.bots_in_range < self.max_bots:
            return
        if c.bots_in_range > self.max_bots:
            # print("Updating min to %d" % c.bots_in_range)
            self.max_bots = c.bots_in_range
            self.max_position = c.center
        elif c.center.distance_to_origin() < self.max_position.distance_to_origin():
            self.max_position = c.center
            
day23().test('Sample input', [
    'pos=<0,0,0>, r=4',
    'pos=<1,0,0>, r=1',
    'pos=<4,0,0>, r=3',
    'pos=<0,2,0>, r=1',
    'pos=<0,5,0>, r=3',
    'pos=<0,0,3>, r=1',
    'pos=<1,1,1>, r=1',
    'pos=<1,1,2>, r=1',
    'pos=<1,3,1>, r=1',
], '7')

day23().test('Sample input part 2', [
    'pos=<10,12,12>, r=2',
    'pos=<12,14,12>, r=2',
    'pos=<16,12,12>, r=4',
    'pos=<14,14,14>, r=6',
    'pos=<50,50,50>, r=200',
    'pos=<10,10,10>, r=5',
], '6', '36')

day23().solve()

File: 2018/day23.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>

#define MAX_BOTS 1000
#define SPLIT_PARTS 8

typedef struct {
    int x;
    int y;
    int z;
} position;

typedef struct {
    position pos;
    int r;
} nanobot;

typedef struct {
    position center;
    int r;
    int bots_in_range;
} cube;

static int total_bots;
static nanobot bots[MAX_BOTS];

static struct {
    int max_bots;
    position max_pos;
} search;

int distance_between(const position *a, const position *b)
{
    return abs(a->x - b->x) + abs(a->y - b->y) + abs(a->z - b->z);
}

int distance_to_origin(const position *p)
{
    return abs(p->x) + abs(p->y) + abs(p->z);
}

void position_translate(position *p, int dx, int dy, int dz)
{
    p->x += dx;
    p->y += dy;
    p->z += dz;
}

int nanobot_in_range(const nanobot *bot, const position *pos)
{
    return distance_between(&bot->pos, pos) <= bot->r;
}

int min_axis_dist(int target, int a_min, int a_max)
{
    if (target < a_min) {
        return a_min - target;
    } else if (target > a_max) {
        return target - a_max;
    } else {
        return 0;
    }
}

int cube_intersects_bot(const cube *c, const nanobot *bot)
{
    if (c->r == 0) {
        return nanobot_in_range(bot, &c->center);
    }
    // Calculate minimum distance required to travel along each axis to
    // arrive at the bounding box. Sum to find the Manhattan distance.
    int distance =
        min_axis_dist(bot->pos.x, c->center.x - c->r, c->center.x + c->r - 1) +
        min_axis_dist(bot->pos.y, c->center.y - c->r, c->center.y + c->r - 1) +
        min_axis_dist(bot->pos.z, c->center.z - c->r, c->center.z + c->r - 1);
    return distance <= bot->r;
}

void init_cube(cube *c)
{
    int in_range = 0;
    for (int i = 0; i < total_bots; i++) {
        if (cube_intersects_bot(c, &bots[i])) {
            in_range++;
        }
    }
    c->bots_in_range = in_range;
    //printf("Cube: bots in range = %d\n", c->bots_in_range);
}

int sort_by_bots(const void *a, const void *b)
{
    const cube *ca = (cube *) a;
    const cube *cb = (cube *) b;
    return cb->bots_in_range - ca->bots_in_range;
}

int split_cube(const cube *c, cube *parts)
{
    if (c->r == 0) {
        return 0;
    }
    int new_r = c->r / 2;
    int low = new_r > 0 ? -new_r : -1;
    int high = new_r;
    for (int i = 0; i < SPLIT_PARTS; i++) {
        parts[i].r = new_r;
        parts[i].center = c->center;
    }
    position_translate(&parts[0].center, low, low, low);
    position_translate(&parts[1].center, low, low, high);
    position_translate(&parts[2].center, low, high, low);
    position_translate(&parts[3].center, low, high, high);
    position_translate(&parts[4].center, high, low, low);
    position_translate(&parts[5].center, high, low, high);
    position_translate(&parts[6].center, high, high, low);
    position_translate(&parts[7].center, high, high, high);
    int valid_parts = 0;
    for (int i = 0; i < SPLIT_PARTS; i++) {
        init_cube(&parts[i]);
        if (parts[i].bots_in_range > 0) {
            valid_parts++;
        }
    }
    qsort(parts, SPLIT_PARTS, sizeof(cube), sort_by_bots);
    return valid_parts;
}

void update_search_state(const cube *c)
{
    if (c->r != 0 || c->bots_in_range < search.max_bots) {
        return;
    }
    if (c->bots_in_range > search.max_bots) {
        // printf("Updating max to %d\n", c->bots_in_range);
        search.max_bots = c->bots_in_range;
        search.max_pos = c->center;
    } else if (distance_to_origin(&c->center) < distance_to_origin(&search.max_pos)) {
        search.max_pos = c->center;
    }
}

void do_search(const cube *cubes, int size)
{
    cube parts[SPLIT_PARTS];
    for (int i = 0; i < size; i++) {
        const cube *c = &cubes[i];
        if (c->bots_in_range > search.max_bots) {
            update_search_state(c);
            int num_parts = split_cube(c, parts);
            do_search(parts, num_parts);
        }
    }
}

int solve2(void)
{
    search.max_bots = 0;
    cube start = {{0, 0, 0}, 1 << 30};
    init_cube(&start);

    do_search(&start, 1);

    return distance_to_origin(&search.max_pos);
}

void read_input(void)
{
    int x, y, z, r;
    while (scanf("pos=<%d,%d,%d>, r=%d\n", &x, &y, &z, &r) == 4) {
        bots[total_bots].pos.x = x;
        bots[total_bots].pos.y = y;
        bots[total_bots].pos.z = z;
        bots[total_bots].r = r;
        total_bots++;
    }
}

int main(int argc, char **argv)
{
    read_input();

    int result = solve2();
    printf("Solution 2: %d\n", result);    

    return 0;
}